File: src/Backend/Infrastructure/Services/AgentMemoryService.cs
using System.Net.Http.Json;
using Application.Common.Interfaces;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace Infrastructure.Services;

public class AgentMemoryService : IAgentMemoryService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<AgentMemoryService> _logger;
    private readonly string _baseUrl;

    public AgentMemoryService(HttpClient httpClient, IConfiguration configuration, ILogger<AgentMemoryService> logger)
    {
        _httpClient = httpClient;
        _logger = logger;
        _baseUrl = configuration["AgentMemory:BaseUrl"] ?? "http://localhost:3111";
    }

    public async Task<bool> SaveMemoryAsync(MemorySaveRequest request, CancellationToken cancellationToken = default)
    {
        try
        {
            var payload = new
            {
                content = request.Content,
                concepts = request.Tags ?? Array.Empty<string>()
            };

            var response = await _httpClient.PostAsJsonAsync($"{_baseUrl}/agentmemory/remember", payload, cancellationToken);
            return response.IsSuccessStatusCode;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to save memory to AgentMemory at {Url}", _baseUrl);
            return false;
        }
    }

    public async Task<IEnumerable<MemoryResult>> RecallMemoryAsync(MemoryRecallRequest request, CancellationToken cancellationToken = default)
    {
        try
        {
            var payload = new { query = request.Query };
            var response = await _httpClient.PostAsJsonAsync($"{_baseUrl}/agentmemory/search", payload, cancellationToken);
            if (!response.IsSuccessStatusCode)
            {
                return Enumerable.Empty<MemoryResult>();
            }

            using var doc = await System.Text.Json.JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync(cancellationToken), cancellationToken: cancellationToken);
            var list = new List<MemoryResult>();

            if (doc.RootElement.TryGetProperty("results", out var resultsElem) && resultsElem.ValueKind == System.Text.Json.JsonValueKind.Array)
            {
                foreach (var item in resultsElem.EnumerateArray())
                {
                    var score = item.TryGetProperty("score", out var s) ? s.GetDouble() : 0.0;
                    var content = request.Query;
                    var id = "mem_" + Guid.NewGuid().ToString("N");

                    if (item.TryGetProperty("observation", out var obs))
                    {
                        if (obs.TryGetProperty("id", out var idProp)) id = idProp.GetString() ?? id;
                        if (obs.TryGetProperty("narrative", out var nProp)) content = nProp.GetString() ?? content;
                        else if (obs.TryGetProperty("facts", out var fProp) && fProp.ValueKind == System.Text.Json.JsonValueKind.Array && fProp.GetArrayLength() > 0)
                        {
                            content = fProp[0].GetString() ?? content;
                        }
                    }

                    list.Add(new MemoryResult(id, content, score, null));
                }
            }

            return list;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to recall memory from AgentMemory at {Url}", _baseUrl);
            return Enumerable.Empty<MemoryResult>();
        }
    }

    public async Task<bool> IsHealthyAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var response = await _httpClient.GetAsync($"{_baseUrl}/agentmemory/health", cancellationToken);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}

File: src/Frontend/src/services/agentMemoryService.ts
export interface MemorySaveRequest {
  content: string;
  tags?: string[];
  category?: string;
}

export interface MemoryRecallRequest {
  query: string;
  limit?: number;
}

export interface MemoryResult {
  id: string;
  content: string;
  score: number;
  tags?: string[];
}

const AGENT_MEMORY_URL = import.meta.env.VITE_AGENT_MEMORY_URL || 'http://localhost:3111';

export const agentMemoryService = {
  async saveMemory(req: MemorySaveRequest): Promise<boolean> {
    try {
      const res = await fetch(`${AGENT_MEMORY_URL}/agentmemory/remember`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          content: req.content,
          concepts: req.tags || [],
        }),
      });
      return res.status === 201 || res.status === 200;
    } catch (err) {
      console.error('Failed to save agent memory:', err);
      return false;
    }
  },

  async recallMemory(query: string, limit = 5): Promise<MemoryResult[]> {
    try {
      const res = await fetch(`${AGENT_MEMORY_URL}/agentmemory/search`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ query, limit }),
      });
      if (!res.ok) return [];
      const data = await res.json();
      const rawResults = data.results || [];
      return rawResults.map((r: any) => ({
        id: r.observation?.id || r.id || 'mem_' + Math.random(),
        content: r.observation?.narrative || r.observation?.facts?.[0] || r.content || query,
        score: r.score || 0,
        tags: r.observation?.concepts || [],
      }));
    } catch (err) {
      console.error('Failed to recall agent memory:', err);
      return [];
    }
  },

  async checkHealth(): Promise<boolean> {
    try {
      const res = await fetch(`${AGENT_MEMORY_URL}/agentmemory/health`);
      return res.ok;
    } catch {
      return false;
    }
  },
};
